protocol/communities/request_to_join.go

```diff
@@ -74,6 +74,19 @@ func (r *RequestToJoin) Empty() bool {
 	return len(r.ID)+len(r.PublicKey)+int(r.Clock)+len(r.ENSName)+len(r.ChatID)+len(r.CommunityID)+int(r.State) == 0
 }
 
+func (r *RequestToJoin) ShouldRetainDeclined(clock uint64) (bool, error) {
+	if r.State != RequestToJoinStateDeclined {
+		return false, nil
+	}
+
+	declineExpiryClock, err := AddTimeoutToRequestToJoinClock(r.Clock)
```

Comment:
This timeout is for admins I'm guessing? For how long they should retain a request to join before the control node validates it?
If so, why not for approvals too?

Reply:
No, this is a cool-down period for declined requests to prevent spam. Once a user is rejected from the community, all their subsequent requests should be automatically declined for a week.

The logic was already there; I just extracted it and, in my opinion, gave it a more precise name: `shouldUserRetainDeclined` over `isUserRejectedFromCommunity`.

Reply:
> Once a user is rejected from the community, all their subsequent requests should be automatically declined for a week.

Do you remember the reason why we only decline for the next 7 days? I thought we always rejected requests after that. 
I'm guessing it's mitigate lost messages or something?

Reply:
> Do you remember the reason why we only decline for the next 7 days? I thought we always rejected requests after that.

I have no idea, the logic to reject new requests if previous one was declined less than a week ago was introduced with this commit: 0eff61c57ae8fa1e80a3032e1ba8ada583598747

@smohamedjavid could you please shed more light on it?

Reply:
I dug through the PR and issues and found this:
> If the request to join a community is ignored or unattended by the Admin/Community Manager for more than seven (7) days, the request should be marked as expired, and the community "Join request" notification from Admin/Community Manager's Activity Center should be deleted.

This makes sense to me. If a request was not replied to in 7 days, the user can request again.

However, the code as it is right now doesn't reflect that. The code seems about declined requests, but maybe that's just the name? It's worth double checking @osmaczko

Reply:
I refactored the code by preserving the logic. I believe fixing the requirements is out of scope for this PR.

Take a look at the code before the refactoring:
https://github.com/status-im/status-go/blob/0eff61c57ae8fa1e80a3032e1ba8ada583598747/protocol/communities/manager.go#L1383-L1389

and after the refactoring:
https://github.com/status-im/status-go/blob/4d0d508f206a97e4f61f7ede8a41e1f48532f352/protocol/communities/manager.go#L2200-L2206

It does exactly the same but with different naming.

Reply:
Take a closer look at the [changes](https://github.com/status-im/status-go/commit/0eff61c57ae8fa1e80a3032e1ba8ada583598747#diff-0c22647108517d38489ae5faa65a914ade3be02ea7a7e37355a82739da9aeb69) to `isUserRejectedFromCommunity` - it can be translated to: "should retain declined"

Reply:
> Take a closer look at the [changes](https://github.com/status-im/status-go/commit/0eff61c57ae8fa1e80a3032e1ba8ada583598747#diff-0c22647108517d38489ae5faa65a914ade3be02ea7a7e37355a82739da9aeb69) to isUserRejectedFromCommunity I mentioned - it can be translated to: "should retain declined"

You are right. The PR looks good to me in terms of the request to join handling.